### PR TITLE
Added Dimension Data examples

### PR DESCRIPTION
**Pull request by trevorflanagan**
Bumped version to 2.2.0-SNAPSHOT across all modules

Added three examples:
1. ```org.jclouds.examples.dimensiondata.cloudcontrol.DeployNetworkDomainVlanAndServer``` - deploys Network Domain, VLAN and a Server and waits for them to be state Normal and in the case of the Server for it also to be started.
2. ```org.jclouds.examples.dimensiondata.cloudcontrol.DeleteServerVlanAndNetworkDomain``` - deletes all of the assets deployed in 1.
3. ```org.jclouds.examples.dimensiondata.cloudcontrol.NetworkDomainTearDown``` - given a Network Domain - clean down all of the associated assets.

### CODE REVIEW

File: dimensiondata/pom.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        <artifactId>jclouds-examples</artifactId>
+        <version>2.2.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>dimensiondata-cloudcontrol-examples</artifactId>
+    <name>dimensiondata-cloudcontrol-examples</name>
+    <version>2.2.0-SNAPSHOT</version>
+
+    <properties>
+        <jclouds.version>2.1.0</jclouds.version>
```
Comment (nacx) [inline dimensiondata/pom.xml]: This should probably pull 2.2.0-snapshot too, to make sure these examples don't have runtime issues.

File: dimensiondata/pom.xml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        <!-- 3rd party dependencies -->
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <version>1.0.13</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.jclouds.driver</groupId>
+            <artifactId>jclouds-slf4j</artifactId>
+            <version>2.2.0-SNAPSHOT</version>
```
Comment (nacx) [inline dimensiondata/pom.xml]: Use `jclouds.version` once changed.

File: dimensiondata/pom.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.1</version>
+                <configuration>
+                    <encoding>${project.build.sourceEncoding}</encoding>
+                    <source>1.6</source>
+                    <target>1.6</target>
```
Comment (nacx) [inline dimensiondata/pom.xml]: Let's move this to 1.7

File: dimensiondata/pom.xml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                </configuration>
+            </plugin>
+            <plugin>
+              <artifactId>maven-assembly-plugin</artifactId>
+              <version>2.2.1</version>
+              <configuration>
+                <descriptors>
+                  <descriptor>src/main/assembly/jar-with-dependencies.xml</descriptor>
```
Comment (nacx) [inline dimensiondata/pom.xml]: Can't you use the default descriptor from the maven assembly plugin? Why do we need a custom one?

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+import org.jclouds.dimensiondata.cloudcontrol.domain.TagKey;
+import org.jclouds.dimensiondata.cloudcontrol.domain.Vlan;
+import org.jclouds.dimensiondata.cloudcontrol.options.DatacenterIdListFilters;
+import org.jclouds.logging.slf4j.config.SLF4JLoggingModule;
+
+import static org.jclouds.examples.dimensiondata.cloudcontrol.WaitForUtils.*;
+
+/**
+ * This class will attempt to delete the assets created in org.jclouds.examples.dimensiondata.cloudcontrol.DeployNetworkDomainVlanAndServer:
+ * <ol>Server</ol>
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java]: Wrong HTML tags. `<ol>` should enclose `<li>` tags.

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+ * <ol>Vlan</ol>
+ * <ol>Network Domain</ol>
+ * <ol>Tag Key</ol>
+ */
+public class DeleteServerVlanAndNetworkDomain
+{
+    private static final String AU_9 = "AU9";
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java]: Better `private static final String REGION = System.getProperty("jclouds.region", "AU9")` to let users customize it?

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                .endpoint(endpoint)
+                .credentials(username, password)
+                .modules(ImmutableSet.<Module>of(new SLF4JLoggingModule()))
+                .buildApi(DimensionDataCloudControlApi.class);
+
+        /*
+         * Retrieve the Guice injector from the context.
+         * We will use this for retrieving the some Predicates that are used by the following operations.
+         */
+        Injector injector = contextBuilder.buildInjector();
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java]: Don't do this, as you're building the context twice. Instead:

```java
ApiContext<DimensionDataCloudControlApi> ctx = contextBuilder
   .endpoint(endpoint)
   .credentials(username, password)
   .modules(ImmutableSet.of(new SLF4JLoggingModule()))
   .build();

Injector injector = ctx.utils().injector();
DimensionDataCloudControlApi api = ctx.getApi();
```

Also remember to always close the `ctx` in a `finally` clause.

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeployNetworkDomainVlanAndServer.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+import org.jclouds.dimensiondata.cloudcontrol.options.DatacenterIdListFilters;
+import org.jclouds.logging.slf4j.config.SLF4JLoggingModule;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.jclouds.examples.dimensiondata.cloudcontrol.WaitForUtils.*;
+
+/**
+ * This class will attempt to Deploy:
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeployNetworkDomainVlanAndServer.java]: Same comments about HTML tags and context creation. Apply everywhere.

**Comment by trevorflanagan** [comment]
Thanks for taking a look @nacx I will make the required changes.

File: dimensiondata/README.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+```
+mvn dependency:copy-dependencies "-DoutputDirectory=./lib"
+```
+
+If you also want to download the source jars, run:
+
+```
+mvn dependency:copy-dependencies "-DoutputDirectory=./lib" "-Dclassifier=sources"
+```
```
Comment (nacx) [inline dimensiondata/README.md]: This looks too complicated. Given that you already have an assembly that produces a jar with all required dependencies, the build and run instructions should simply be:
```bash
mvn package
java -jar target/dimensiondata-cloudcontrol-examples-jar-with-dependencies.jar <apiEndpoint> <username> <password>
```

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+import static org.jclouds.examples.dimensiondata.cloudcontrol.WaitForUtils.*;
+
+/**
+ * This class will attempt to delete the assets created in org.jclouds.examples.dimensiondata.cloudcontrol.DeployNetworkDomainVlanAndServer:
+ * <ul>
+ * <li>Server</li>
+ * <li>Vlan</li>
+ * <li>Network Domain</li>
+ * <li>Tag Key</li>
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java]: End tag `</ul>`

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        String username = args[1];
+        String password = args[2];
+        ApiContext<DimensionDataCloudControlApi> ctx = null;
+        try
+        {
+            ctx = ContextBuilder.newBuilder(DIMENSIONDATA_CLOUDCONTROL_PROVIDER)
+                    .endpoint(endpoint)
+                    .credentials(username, password)
+                    .modules(ImmutableSet.of(new SLF4JLoggingModule()))
+                    .build();
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/DeleteServerVlanAndNetworkDomain.java]: [minor] You could also use try-with-resources to save the finally block:

```java
try (ApiContext<DimensionDataCloudControlApi> ctx =
   ContextBuilder.newBuilder(DIMENSIONDATA_CLOUDCONTROL_PROVIDER)
      .endpoint(endpoint)
      .credentials(username, password)
      .modules(ImmutableSet.of(new SLF4JLoggingModule()))
      .build())
{
...
}
```

File: dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/NetworkDomainTearDown.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+/**
+ * This example shows how a Network Domain and all of it's associated assets are removed.
+ * Takes 4 Program Arguments:
+ * <ul>
+ * <ol>1 Endpoint URL</ol>
+ * <ol>2 Usernamme</ol>
+ * <ol>3 Password</ol>
+ * <ol>4 Network Domain Id</ol>
+ * </ul>
```
Comment (nacx) [inline dimensiondata/src/main/java/org/jclouds/examples/dimensiondata/cloudcontrol/NetworkDomainTearDown.java]: This should be
```html
<ol>
<li></li>
<li></li>
</ol>
```

File: pom.xml [hunk 13]
```diff
@@ -21,7 +21,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>org.apache.jclouds.examples</groupId>
   <artifactId>jclouds-examples</artifactId>
-  <version>2.1.0</version>
+  <version>2.2.0-SNAPSHOT</version>
```
Comment (nacx) [inline pom.xml]: Even if dimension data (in labs) uses the snapshot, keep this version as-is, please.

Comment (trevorflanagan) [reply pom.xml]: Great thanks I wasnt sure what the correct approach was for the parent pom.

**Comment by nacx** [comment]
merged to master as a145a8fd[](http://git-wip-us.apache.org/repos/asf/jclouds-examples/commit/a145a8fd). Thanks, @trevorflanagan!

**Pull request closed without merging**